<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-A01793506/blob/main/SVD_Equipo13_v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Maestría en Inteligencia Artificial Aplicada**
##**Curso: Inteligencia Artificial y Aprendizaje Automático**
###Tecnológico de Monterrey
###Prof Luis Eduardo Falcón Morales

## **Actividad 11 **
###**Reduccion de dimensionalidad**
**Nombres y matrículas de los integrantes del equipo:**

*   Rodrigo Rodriguez Rodriguez A01183284
*   Christian Emilio Saldana Lopez A00506509
*   Irvin Gomez Esquivel A00361034
*   Erick Najera Olivero A01793506

In [ ]:
import numpy as np
import pandas as pd

from sklearn.decomposition import TruncatedSVD

### Ejercicio 1

In [ ]:
#Cargamos los conjuntos de datos necesarios en dataframes
restaurantes = pd.read_csv("geoplaces2.csv", header=0
                           , sep=",", encoding='latin-1')

rating=pd.read_csv("rating_final.csv", header=0, sep=",", encoding='latin-1')



In [ ]:
#Creamos el DF dfestudio con las columnas del conjunto de datos e rating, visualizamos el tamaño y los registros de DF
dfestudio=pd.DataFrame(rating)
dfestudio=dfestudio[['userID','placeID','rating']]
print(dfestudio.columns)
print(dfestudio.shape)
print (dfestudio.head)

Index(['userID', 'placeID', 'rating'], dtype='object')
(1161, 3)
<bound method NDFrame.head of      userID  placeID  rating
0     U1077   135085       2
1     U1077   135038       2
2     U1077   132825       2
3     U1077   135060       1
4     U1068   135104       1
...     ...      ...     ...
1156  U1043   132630       1
1157  U1011   132715       1
1158  U1068   132733       1
1159  U1068   132594       1
1160  U1068   132660       0

[1161 rows x 3 columns]>


In [ ]:
dfestudio

userID  placeID  rating
0     U1077   135085       2
1     U1077   135038       2
2     U1077   132825       2
3     U1077   135060       1
4     U1068   135104       1
...     ...      ...     ...
1156  U1043   132630       1
1157  U1011   132715       1
1158  U1068   132733       1
1159  U1068   132594       1
1160  U1068   132660       0

[1161 rows x 3 columns]

In [ ]:
#Visualizamos la informacion del DF restarurantes
restaurantes_nombres = restaurantes[['placeID', 'name']]    # DataFrame
restaurantes_nombres.head()

placeID                            name
0   134999                 Kiku Cuernavaca
1   132825                 puesto de tacos
2   135106      El Rincón de San Francisco
3   132667  little pizza Emilio Portes Gil
4   132613                   carnitas_mata

In [ ]:
#Conjuntamos la información de ambos DF a traves de la información de la columna placeID:
tabla_relacionalEstudio1 = pd.merge(dfestudio,restaurantes_nombres, on='placeID')   # dataFrame
#Visualizamos los primeros 5 registros
tabla_relacionalEstudio1.head()

userID  placeID  rating                    name
0  U1077   135085       2  Tortas Locas Hipocampo
1  U1108   135085       1  Tortas Locas Hipocampo
2  U1081   135085       1  Tortas Locas Hipocampo
3  U1056   135085       2  Tortas Locas Hipocampo
4  U1134   135085       2  Tortas Locas Hipocampo

### Ejercicio2

In [ ]:
#Generamos la matriz de utilidad que proporciona la evaluación que cada usuario asigna a cada restaurante, se rellena con 0 las celdas sin registro
UtMx = tabla_relacionalEstudio1.pivot_table(values='rating', index='userID', columns='name', fill_value=0)

UtMx.head(10)

name    Abondance Restaurante Bar  Arrachela Grill  Cabana Huasteca  \
userID                                                                
U1001                           0                0                0   
U1002                           0                0                0   
U1003                           0                0                2   
U1004                           0                0                0   
U1005                           0                0                0   
U1006                           0                0                0   
U1007                           0                0                0   
U1008                           1                0                0   
U1009                           0                0                1   
U1010                           0                0                0   

name    Cafe Chaires  Cafeteria cenidet  Cafeteria y Restaurant El Pacifico  \
userID                                                                        
U1001              0                  0                                   0   
U1002              0                  0                                   0   
U1003              0                  0                                   0   
U1004              0                  0                                   2   
U1005              0                  0                                   2   
U1006              1                  0                                   0   
U1007              1                  0                                   1   
U1008              0                  0                                   0   
U1009              0                  0                                   0   
U1010              0                  0                                   0   

name    Carls Jr  Carnitas Mata  Calle 16 de Septiembre  \
userID                                                    
U1001          0                                      0   
U1002          0                                      0   
U1003          0                                      0   
U1004          0                                      0   
U1005          0                                      0   
U1006          0                                      0   
U1007          0                                      0   
U1008          0                                      0   
U1009          0                                      0   
U1010          0                                      0   

name    Carreton de Flautas y Migadas  Cenaduria El Rincón de Tlaquepaque  \
userID                                                                       
U1001                               0                                    0   
U1002                               0                                    0   
U1003                               0                                    0   
U1004                               0                                    0   
U1005                               0                                    0   
U1006                               0                                    0   
U1007                               0                                    0   
U1008                               0                                    0   
U1009                               0                                    0   
U1010                               0                                    0   

name    ...  puesto de tacos  rockabilly   shi ro ie  sirloin stockade  \
userID  ...                                                              
U1001   ...                2            0          0                 0   
U1002   ...                2            0          0                 0   
U1003   ...                2            2          0                 0   
U1004   ...                0            0          0                 0   
U1005   ...                0            0          0                 0   
U1006   ...                1      

In [ ]:
#Verificamos dimensiones de la matriz de utilidad
UtMx.shape

(138, 129)

In [ ]:
print('Total de elementos de la matriz de utilidad: %d' % (UtMx.size))
print('Total de elementos diferentes de cero: %d' % (np.count_nonzero(UtMx)))
print('Porcentaje de elementos diferentes de cero: %.1f%%' % (100 * np.count_nonzero(UtMx) / UtMx.size))

Total de elementos de la matriz de utilidad: 17802
Total de elementos diferentes de cero: 907
Porcentaje de elementos diferentes de cero: 5.1%


In [ ]:
X = UtMx.T     # Tomamos la traspuesta para que los renglones sean los restaurantes y las columnas los usuarios.
X.shape

(129, 138)

In [ ]:
# Obtengamos la cantidad de información contenida en la factorización SVD, considerando
# todos los componentes posibles, es decir, todos los valores singulares

 # El número de componentes debe ser menor que el número de características (features/usuarios), que son 138:
SVD = TruncatedSVD(n_components=137, random_state=1)
SVD.fit(X)

num_sv =10 # En particular los 10 primeros vectores y valores singulares de mayor magnitud tienen esta cantidad de información.

print('Cantidad de información simplificada con los primeros %d vectores singulares:' % num_sv)
print('%.1f%%' %  (100 * (1- (SVD.singular_values_[0:num_sv]).sum() / (SVD.singular_values_).sum())))

Cantidad de información simplificada con los primeros 10 vectores singulares:
75.6%


In [ ]:
# Usamos el total de componentes disponibles 138

num_sv = 137 ### para este primer ejercicio usemos todas

SVDT = TruncatedSVD(n_components=num_sv, random_state=1)

resultant_matrix = SVDT.fit_transform(X)
print('Cantidad de  valores singulares obtenidos:' , round(len(SVDT.singular_values_),2))
resultant_matrix.shape

Cantidad de  valores singulares obtenidos: 129


(129, 129)

In [ ]:
# Obtengamos la matriz de correlación de Pearson entre todas las
# variables latentes de los restaurantes:

corrMx = np.corrcoef(resultant_matrix, )
corrMx.shape

(129, 129)

### Ejercicio 3

In [ ]:
#Revisamos la evaluacion de Subway obteniendo el id dentro de la matriz de utilidad
me_gusta = "Subway"

names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

id_megusta

88

In [ ]:
#Visualizamos la correlación de "Subway" con todos los demas restaurantes en la matriz
corr_recomienda = corrMx[id_megusta]
print(corr_recomienda.shape)
print(corr_recomienda)

(129,)
[-7.69028369e-04 -7.87822168e-05 -5.27330975e-03  8.58511349e-04
 -3.03891148e-03  3.97495832e-05 -1.54820208e-03 -5.29347145e-03
  7.27520383e-04 -1.55098452e-03 -1.84483094e-03  3.01090491e-01
 -3.06938616e-05 -3.66639157e-03 -2.28024815e-03 -1.70971180e-03
  2.66961340e-03  4.58467365e-04 -2.11449533e-03 -6.20445134e-03
  3.34268563e-03  1.16598607e-01  1.87080102e-03  2.53682556e-03
 -5.48801141e-03 -2.91288266e-03  2.02825342e-03 -1.36607265e-03
 -2.01695899e-03  1.12058244e-01 -2.36235727e-03  6.52159239e-04
  3.53495770e-03 -3.62793330e-03  1.69065407e-03 -4.51837401e-03
 -2.09301342e-03  2.34634298e-01 -3.80413494e-03 -5.11845710e-04
 -2.27148704e-03  3.88178002e-01 -2.07427108e-03 -2.29428835e-03
  1.32565455e-03  3.01002479e-01  1.71433518e-01 -3.29715354e-03
 -4.74480596e-03 -8.02527884e-04 -2.65071452e-03  1.12018889e-01
 -3.91080724e-03 -2.06369812e-03  1.16008629e-01  2.27855804e-01
  4.28342495e-04 -8.69739761e-04  6.09520875e-01 -6.43364064e-03
  4.25467169e-01 -

In [ ]:
#Correlacion del restaurante consigo mismo, se visualiza para estudio y validación que la correlacion con el mismo debe sr igual a 1 o muy proximo a este
corr_recomienda[id_megusta]


1.0

In [ ]:
#Obtenemos las 10 mejores recomendaciones del modelo de acuerdo a valoraciones buenas a "Subway" y agregamos el valor

Best10=list(names[(corr_recomienda > .23) & (corr_recomienda < 0.99999)])

for pos in Best10:
 id_otra = names_list.index(pos)
 print('Restaurante',pos, ' | valor correlación: ', round(corr_recomienda[id_otra],4))



Restaurante Chilis Cuernavaca  | valor correlación:  0.3011
Restaurante Log Yin  | valor correlación:  0.2346
Restaurante Mariscos Tia Licha  | valor correlación:  0.3882
Restaurante Mikasa  | valor correlación:  0.301
Restaurante Restaurant Teely  | valor correlación:  0.6095
Restaurante Restaurant and Bar and Clothesline Carlos N Charlies  | valor correlación:  0.4255
Restaurante Restaurant los Pinos  | valor correlación:  0.2851
Restaurante Rincon del Bife  | valor correlación:  0.4298
Restaurante Sanborns Casa Piedra  | valor correlación:  0.4172
Restaurante Vips  | valor correlación:  0.5063


### Ejercicio4
Busquemosel mejor numerdo de componentes y cuantos usar

In [ ]:
num_sv =11 #reduscamos la factorizacion a 11 componentes

SVD = TruncatedSVD(n_components=num_sv, random_state=1)

resultant_matrix4 = SVD.fit_transform(X)
print('Dimensiones resultante:', resultant_matrix4.shape, '\n')

#Obtenemos la matriz de correlación 
corrMx4 = np.corrcoef(resultant_matrix4, )

me_gusta = "Subway"
names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

#Correlación del restaurante con los demas
corr_recomienda4 = corrMx4[id_megusta]

print('Id del resaturante en la matriz: ',id_megusta,'\n')
corr_recomienda4[id_megusta]
mejores10=list(names[(corr_recomienda4 > .996) & (corr_recomienda4 < 0.9999)])  
print('10 mejores recomendaciones\n')
for pos in mejores10:
 id_otra = names_list.index(pos)
 print('Restaurante',pos, ' | valor correlación: ', round(corr_recomienda4[id_otra],4))

Dimensiones resultante: (129, 11) 

Id del resaturante en la matriz:  88 

10 mejores recomendaciones

Restaurante Chilis Cuernavaca  | valor correlación:  0.9991
Restaurante Log Yin  | valor correlación:  0.9969
Restaurante Mariscos Tia Licha  | valor correlación:  0.9995
Restaurante Mikasa  | valor correlación:  0.9986
Restaurante Paniroles  | valor correlación:  0.9998
Restaurante Restaurant and Bar and Clothesline Carlos N Charlies  | valor correlación:  0.9993
Restaurante Restaurant los Pinos  | valor correlación:  0.9967
Restaurante Rincon del Bife  | valor correlación:  0.9997
Restaurante Sanborns Casa Piedra  | valor correlación:  0.997
Restaurante Vips  | valor correlación:  0.9978


In [ ]:
#Obtenemos el numero de recomendaciones iguales respecto a los 10 mejores con todos los componentes
#Se obtiene el mismo resutlado con solo 10 componentes
print("Numero de Recomendaciones iguales:",len(set(mejores10).intersection(Best10)),'\n')

print('Cantidad de información simplificada con los primeros %d vectores singulares:' % num_sv)
print('%.1f%%' %  (100 * (1- (SVDT.singular_values_[0:num_sv]).sum() / (SVDT.singular_values_).sum())))


Numero de Recomendaciones iguales: 9 

Cantidad de información simplificada con los primeros 11 vectores singulares:
73.8%


### Ejercicio 5

In [ ]:
#Tomamos el valor de 10 para los componentes en al factorizacion
num_sv = 10

SVD = TruncatedSVD(n_components=num_sv, random_state=1)

resultant_matrix5 = SVD.fit_transform(X)
resultant_matrix5.shape

(129, 10)

In [ ]:
#Realizamos nuevamente el ejercicio con el "restaurante tacos de barbacoa enfrente del Tec"
me_gusta = "tacos de barbacoa enfrente del Tec"

#Obtenemos la matriz de correlación 
corrMx5 = np.corrcoef(resultant_matrix5, )

names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

print('Id del restaurante: ',id_megusta)

Id del restaurante:  124


In [ ]:
corr_recomienda5 = corrMx5[id_megusta]
print(corr_recomienda5.shape)
print(corr_recomienda5)

(129,)
[ 8.26808339e-02 -1.60487509e-01  5.32570598e-01 -2.97836619e-01
 -4.73811117e-03 -7.41763626e-02  1.20101547e-01  9.66995956e-01
  9.97711045e-01  2.94148153e-01  9.44135223e-02  3.02493721e-02
  2.81438921e-01  3.07379599e-02  1.63596498e-01 -4.50130527e-02
 -3.79465414e-02  1.89886821e-01 -1.33824127e-01  2.77343149e-01
 -7.28129478e-02 -3.28427076e-02  3.18813635e-01  2.90548961e-01
  9.95809317e-01  9.98270371e-01  2.86082505e-01  1.80550369e-01
  2.93887446e-01 -3.40720748e-02  5.85428853e-01  1.36200061e-01
  3.83557567e-01  4.90786788e-01 -7.04947453e-02 -4.84781779e-01
  9.98219729e-01  6.17595183e-03 -2.20939293e-02  1.33496375e-01
 -1.29114556e-01  2.79102713e-02 -2.59011076e-03  2.59717547e-01
  3.89663781e-01  2.71522358e-02  4.70112294e-02  2.90859259e-01
  9.99520175e-01  2.94368156e-01 -2.06162660e-01 -2.22211392e-02
  4.26175353e-01  1.06394830e-01 -3.07773168e-02  6.00756809e-04
  8.77004770e-02 -3.02227119e-01  4.95320763e-02 -1.76066193e-01
  3.00827553e-02 -

In [ ]:
#Correlacion del restaurante consigo mismo, se visualiza para estudio y validación que la correlacion con el mismo debe sr igual a 1 o muy proximo a este
corr_recomienda5[id_megusta]

1.0

In [ ]:
#Obtenemos las 10 mejores recomendaciones de acuerdo a valoracion del restaurante.
mejores10T= list(names[(corr_recomienda5 > .998) & (corr_recomienda5 < 0.9999999999)])
print('10 mejores recomendaciones\n')
for pos in mejores10T:
 id_otra = names_list.index(pos)
 print('Restaurante',pos, ' | valor correlación: ', round(corr_recomienda5[id_otra],4))

10 mejores recomendaciones

Restaurante Hamburguesas La perica  | valor correlación:  0.9983
Restaurante Little Cesarz  | valor correlación:  0.9982
Restaurante Pollo_Frito_Buenos_Aires  | valor correlación:  0.9995
Restaurante Taqueria EL amigo   | valor correlación:  0.9999
Restaurante carnitas_mata  | valor correlación:  0.9994
Restaurante little pizza Emilio Portes Gil  | valor correlación:  0.9997
Restaurante palomo tec  | valor correlación:  0.9994
Restaurante puesto de gorditas  | valor correlación:  0.9994
Restaurante tacos abi  | valor correlación:  1.0
Restaurante vips  | valor correlación:  0.9998


###Conclusiones:

En la práctica de aplicación de SVD para reducción de la dimensionalidad podemos observar que al reducir la matriz original de información a una de menor dimensión, obtenemos los valores singulares que representan la cantidad de información, en el ejercicio vimos que al usar solo 10 valores singulares, logramos reducir la cantidad de información en un 76%, esta contiene, para este caso, el grado de satisfacción de los clientes de los restaurantes mencionados, es decir, donde la valoración fue similar, por lo tanto las características dentro de esta asociación representan valores latentes. Pudimos observar en el ejercicio que para el restaurante "tacos de barbacoa enfrente del Tec" se tiene una valoración muy cercana para establecimientos del mismo tipo de comida, por ejemplo, vemos con alta correlación, restaurante Vips con 0.99; tacos abi con 1.0; Carnitas mata con 0.99  y taquería el amigo con 0.99. Teniendo en común que son establecimientos de comida típica Mexicana, por lo que concluimos que el sistema de recomendación está basado en que los usuarios valoraron igual a los restaurantes de acuerdo al mismo tipo de comida.

### Ejercicio 6

In [ ]:
#Cargamos el nuevo dataframe con la informacion de la puntuación dada a las comidas por los usuaarios de cada restaurante
dfestudio2=rating[['userID','placeID','food_rating']]
#Conjuntamos la informacio del dataframe con el dataframe de los nombres de los restaurantes
tabla_relacionalEstudio2 = pd.merge(dfestudio2,restaurantes_nombres, on='placeID')   # dataFrame
#Obtenemos las 5 primeras filas
tabla_relacionalEstudio2.head()

userID  placeID  food_rating                    name
0  U1077   135085            2  Tortas Locas Hipocampo
1  U1108   135085            2  Tortas Locas Hipocampo
2  U1081   135085            2  Tortas Locas Hipocampo
3  U1056   135085            2  Tortas Locas Hipocampo
4  U1134   135085            1  Tortas Locas Hipocampo

### Ejercicio 6.2

In [ ]:
#Generamos la matriz de utilidad mediante el campo UserID, rellando la falta de valores con 0
UtMx = tabla_relacionalEstudio2.pivot_table(values='food_rating', index='userID', columns='name', fill_value=0)

UtMx.head(10)

name    Abondance Restaurante Bar  Arrachela Grill  Cabana Huasteca  \
userID                                                                
U1001                           0                0                0   
U1002                           0                0                0   
U1003                           0                0                2   
U1004                           0                0                0   
U1005                           0                0                0   
U1006                           0                0                0   
U1007                           0                0                0   
U1008                           1                0                0   
U1009                           0                0                2   
U1010                           0                0                0   

name    Cafe Chaires  Cafeteria cenidet  Cafeteria y Restaurant El Pacifico  \
userID                                                                        
U1001              0                  0                                   0   
U1002              0                  0                                   0   
U1003              0                  0                                   0   
U1004              0                  0                                   2   
U1005              0                  0                                   1   
U1006              0                  0                                   0   
U1007              1                  0                                   1   
U1008              0                  0                                   0   
U1009              0                  0                                   0   
U1010              0                  0                                   0   

name    Carls Jr  Carnitas Mata  Calle 16 de Septiembre  \
userID                                                    
U1001          0                                      0   
U1002          0                                      0   
U1003          0                                      0   
U1004          0                                      0   
U1005          0                                      0   
U1006          0                                      0   
U1007          0                                      0   
U1008          0                                      0   
U1009          0                                      0   
U1010          0                                      0   

name    Carreton de Flautas y Migadas  Cenaduria El Rincón de Tlaquepaque  \
userID                                                                       
U1001                               0                                    0   
U1002                               0                                    0   
U1003                               0                                    0   
U1004                               0                                    0   
U1005                               0                                    0   
U1006                               0                                    0   
U1007                               0                                    0   
U1008                               0                                    0   
U1009                               0                                    0   
U1010                               0                                    0   

name    ...  puesto de tacos  rockabilly   shi ro ie  sirloin stockade  \
userID  ...                                                              
U1001   ...                2            0          0                 0   
U1002   ...                2            0          0                 0   
U1003   ...                2            2          0                 0   
U1004   ...                0            0          0                 0   
U1005   ...                0            0          0                 0   
U1006   ...                1      

In [ ]:
UtMx.shape

(138, 129)

In [ ]:
print('Total de elementos de la matriz de utilidad: %d' % (UtMx.size))
print('Total de elementos diferentes de cero: %d' % (np.count_nonzero(UtMx)))
print('Porcentaje de elementos diferentes de cero: %.1f%%' % (100 * np.count_nonzero(UtMx) / UtMx.size))

Total de elementos de la matriz de utilidad: 17802
Total de elementos diferentes de cero: 895
Porcentaje de elementos diferentes de cero: 5.0%


In [ ]:
X = UtMx.T     # Tomamos la traspuesta para que los renglones sean los restaurantes y las columnas los usuarios.
X.shape

(129, 138)

In [ ]:
# Obtengamos la cantidad de información contenida en la factorización SVD, considerando
# los valores singulares más representativos de dicha factorización.
# Para este caso inicial usaremos todos los vectores singulares, para luego
# compararlo con el caso al reducir su dimensión.

 # El número de componentes debe ser menor que el número de características (features/usuarios), que son 138:
SVD = TruncatedSVD(n_components=137, random_state=1)
SVD.fit(X)

num_sv =10 # En particular los 10 primeros vectores y valores singulares de mayor magnitud tienen esta cantidad de información.

print('Cantidad de información simplificada con los primeros %d vectores singulares:' % num_sv)
print('%.1f%%' %  (100 * (1- (SVD.singular_values_[0:num_sv]).sum() / (SVD.singular_values_).sum())))

Cantidad de información simplificada con los primeros 10 vectores singulares:
75.9%


In [ ]:
num_sv = 137 # para este ejercicio usaremos todos

SVD = TruncatedSVD(n_components=num_sv, random_state=1)

resultant_matrix = SVD.fit_transform(X)
resultant_matrix.shape

(129, 129)

In [ ]:
corrMx = np.corrcoef(resultant_matrix, )
corrMx.shape

(129, 129)

### Ejercico 6.3

In [ ]:
#Tomamos el restaurante "subway" para valorar la correlación de este en cuanto a la valoración de la comida con los demas establecimientos
me_gusta = "Subway"

names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

id_megusta

88

In [ ]:
corr_recomienda = corrMx[id_megusta]
print(corr_recomienda.shape)
print(corr_recomienda)

(129,)
[-5.06418108e-03  1.53891772e-03 -7.15705294e-03 -7.51437349e-04
  3.57844065e-04 -4.27908953e-03 -4.10491062e-03 -8.72391553e-03
 -5.80659154e-03 -7.37551348e-03 -6.29677017e-03  1.19529006e-03
 -5.49056441e-03 -6.25106315e-03 -2.26808294e-04  2.26784575e-03
 -4.97923491e-03 -7.33491870e-03 -5.49091909e-05 -6.54487708e-03
 -7.17291384e-04  2.31333734e-01 -9.42653806e-04  1.15809414e-03
 -2.44343966e-03 -2.46720616e-03 -8.43368969e-04 -2.05629573e-03
 -3.70288596e-03  2.50357400e-01 -5.28927854e-03  2.72981977e-03
  4.14365709e-04 -9.51153872e-03  2.32009229e-03 -9.29532082e-03
 -4.03970901e-03  1.31858098e-01 -8.36645261e-03 -2.30125601e-03
 -2.61409620e-03  4.15222053e-01  4.25538306e-03  3.02880631e-03
 -2.46282020e-03  2.09493113e-01  3.12082362e-01 -3.01154158e-03
 -3.10972278e-04 -1.11824996e-03  8.07121501e-04  2.24213782e-01
  2.65096107e-04  6.44807197e-04  2.32708368e-01  3.31706780e-01
  6.95063851e-04  3.47994174e-04  8.16413374e-01  1.04699730e-04
  4.68949686e-01  

In [ ]:
corr_recomienda[id_megusta]

1.0

In [ ]:
Best10=list(names[(corr_recomienda > .232) & (corr_recomienda < 0.99999)])
print('Mejores 10 restaurantes de acuerdo a valoracion de la comida de acuerdo a la correlación con "Subway"\n')
for pos in Best10:
 id_otra = names_list.index(pos)
 print('Restaurante',pos, ' | valor correlación: ', round(corr_recomienda[id_otra],2))

Mejores 10 restaurantes de acuerdo a valoracion de la comida de acuerdo a la correlación con "Subway"

Restaurante Kiku Cuernavaca  | valor correlación:  0.25
Restaurante Mariscos Tia Licha  | valor correlación:  0.42
Restaurante Paniroles  | valor correlación:  0.31
Restaurante Restaurant Familiar El Chino  | valor correlación:  0.23
Restaurante Restaurant Las Mananitas  | valor correlación:  0.33
Restaurante Restaurant Teely  | valor correlación:  0.82
Restaurante Restaurant and Bar and Clothesline Carlos N Charlies  | valor correlación:  0.47
Restaurante Rincon del Bife  | valor correlación:  0.38
Restaurante Sanborns Casa Piedra  | valor correlación:  0.37
Restaurante Vips  | valor correlación:  0.44


### Ejercicio 6.4

In [ ]:
num_sv = 12

SVD = TruncatedSVD(n_components=num_sv, random_state=1)

resultant_matrix6 = SVD.fit_transform(X)
resultant_matrix6.shape

corrMx6 = np.corrcoef(resultant_matrix6, )
corrMx6.shape

me_gusta = "Subway"
names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

corr_recomienda6 = corrMx6[id_megusta]

corr_recomienda6[id_megusta]
mejores10=list(names[(corr_recomienda6 > .39) & (corr_recomienda6 < 0.99999)])
for pos in mejores10:
 id_otra = names_list.index(pos)
 print('Restaurante',pos, ' | valor correlación: ', round(corr_recomienda6[id_otra],4))

Restaurante Arrachela Grill  | valor correlación:  0.7002
Restaurante Cafeteria cenidet  | valor correlación:  0.7622
Restaurante Chilis Cuernavaca  | valor correlación:  0.5368
Restaurante El Oceano Dorado  | valor correlación:  0.8304
Restaurante El cotorreo  | valor correlación:  0.9973
Restaurante Kiku Cuernavaca  | valor correlación:  0.3905
Restaurante Log Yin  | valor correlación:  0.8466
Restaurante Mariscos Tia Licha  | valor correlación:  0.865
Restaurante McDonalds Centro  | valor correlación:  0.7885
Restaurante Mikasa  | valor correlación:  0.9254
Restaurante Paniroles  | valor correlación:  0.8647
Restaurante Restaurant Bar Coty y Pablo  | valor correlación:  0.5656
Restaurante Restaurant Las Mananitas  | valor correlación:  0.8738
Restaurante Restaurant Teely  | valor correlación:  0.9528
Restaurante Restaurant and Bar and Clothesline Carlos N Charlies  | valor correlación:  0.9416
Restaurante Restaurant los Pinos  | valor correlación:  0.8925
Restaurante Rincon del Bife

In [ ]:
print("Numero de Recomandaciones iguales:",len(set(mejores10).intersection(Best10)))

Numero de Recomandaciones iguales: 9


###Conclusiones

Después de realizado el ejercicio aplicando técnica de reducción de dimensionalidad para obtener recomendaciones de acuerdo a la valoración de la comida en los diferentes restaurantes, podemos observar que no existe mucha correlación entre los diferentes restaurantes en términos de la valoración indicada, podemos ver que al aplicar la técnica con todos los componentes obtenemos el ranking de los 10 mejores, donde solo el restaurante Teely tiene un coeficiente de correlación del 0.82 siendo el más alto, seguido de “Restaurant and Bar and Clothesline Carlos N Charlies” con un valor de 0.47 en la correlación, los que lo siguen tienen valores inferiores, por lo tanto al aplicar la técnica SVD encontrando un numero óptimo de 12 vectores singulares se logró obtener entre los 10 mejores 7 restaurantes que se obtuvieron con todos los componentes.

### Ejercicio 7

In [ ]:
#Cargamos el nuevo dataframe con la informacion de la puntuación dada al servicio por los usuarios de cada restaurante
dfestudio3=rating[['userID','placeID','service_rating']]
#Conjuntamos la informacio del dataframe con el dataframe de los nombres de los restaurantes
tabla_relacionalEstudio3 = pd.merge(dfestudio3,restaurantes_nombres, on='placeID')   # dataFrame
#Obtenemos las 5 primeras filas
tabla_relacionalEstudio3.head()

userID  placeID  service_rating                    name
0  U1077   135085               2  Tortas Locas Hipocampo
1  U1108   135085               1  Tortas Locas Hipocampo
2  U1081   135085               1  Tortas Locas Hipocampo
3  U1056   135085               2  Tortas Locas Hipocampo
4  U1134   135085               2  Tortas Locas Hipocampo

### Ejercico 7.2

In [ ]:
#Generamos la matriz de utilidad mediante el campo UserID, rellando la falta de valores con 0
UtMx = tabla_relacionalEstudio3.pivot_table(values='service_rating', index='userID', columns='name', fill_value=0)

UtMx.head(10)

name    Abondance Restaurante Bar  Arrachela Grill  Cabana Huasteca  \
userID                                                                
U1001                           0                0                0   
U1002                           0                0                0   
U1003                           0                0                2   
U1004                           0                0                0   
U1005                           0                0                0   
U1006                           0                0                0   
U1007                           0                0                0   
U1008                           1                0                0   
U1009                           0                0                1   
U1010                           0                0                0   

name    Cafe Chaires  Cafeteria cenidet  Cafeteria y Restaurant El Pacifico  \
userID                                                                        
U1001              0                  0                                   0   
U1002              0                  0                                   0   
U1003              0                  0                                   0   
U1004              0                  0                                   2   
U1005              0                  0                                   2   
U1006              1                  0                                   0   
U1007              1                  0                                   0   
U1008              0                  0                                   0   
U1009              0                  0                                   0   
U1010              0                  0                                   0   

name    Carls Jr  Carnitas Mata  Calle 16 de Septiembre  \
userID                                                    
U1001          0                                      0   
U1002          0                                      0   
U1003          0                                      0   
U1004          0                                      0   
U1005          0                                      0   
U1006          0                                      0   
U1007          0                                      0   
U1008          0                                      0   
U1009          0                                      0   
U1010          0                                      0   

name    Carreton de Flautas y Migadas  Cenaduria El Rincón de Tlaquepaque  \
userID                                                                       
U1001                               0                                    0   
U1002                               0                                    0   
U1003                               0                                    0   
U1004                               0                                    0   
U1005                               0                                    0   
U1006                               0                                    0   
U1007                               0                                    0   
U1008                               0                                    0   
U1009                               0                                    0   
U1010                               0                                    0   

name    ...  puesto de tacos  rockabilly   shi ro ie  sirloin stockade  \
userID  ...                                                              
U1001   ...                1            0          0                 0   
U1002   ...                0            0          0                 0   
U1003   ...                1            1          0                 0   
U1004   ...                0            0          0                 0   
U1005   ...                0            0          0                 0   
U1006   ...                1      

In [ ]:
UtMx.shape

(138, 129)

In [ ]:
print('Total de elementos de la matriz de utilidad: %d' % (UtMx.size))
print('Total de elementos diferentes de cero: %d' % (np.count_nonzero(UtMx)))
print('Porcentaje de elementos diferentes de cero: %.1f%%' % (100 * np.count_nonzero(UtMx) / UtMx.size))

Total de elementos de la matriz de utilidad: 17802
Total de elementos diferentes de cero: 846
Porcentaje de elementos diferentes de cero: 4.8%


In [ ]:
X = UtMx.T     # Tomamos la traspuesta para que los renglones sean los restaurantes y las columnas los usuarios.
X.shape

(129, 138)

In [ ]:
# Obtengamos la cantidad de información contenida en la factorización SVD, considerando
# los valores singulares más representativos de dicha factorización.
# Para este caso inicial usaremos todos los vectores singulares, para luego
# compararlo con el caso al reducir su dimensión.

 # El número de componentes debe ser menor que el número de características (features/usuarios), que son 138:
SVD = TruncatedSVD(n_components=137, random_state=1)
SVD.fit(X)

num_sv =10 # En particular los 10 primeros vectores y valores singulares de mayor magnitud tienen esta cantidad de información.

print('Cantidad de información simplificada con los primeros %d vectores singulares:' % num_sv)
print('%.1f%%' %  (100 * (1- (SVD.singular_values_[0:num_sv]).sum() / (SVD.singular_values_).sum())))

Cantidad de información simplificada con los primeros 10 vectores singulares:
75.6%


In [ ]:
num_sv = 137

SVD = TruncatedSVD(n_components=num_sv, random_state=1)

resultant_matrix = SVD.fit_transform(X)
resultant_matrix.shape

(129, 129)

In [ ]:
corrMx = np.corrcoef(resultant_matrix, )
corrMx.shape

(129, 129)

### Ejercicio 7.3

In [ ]:
#Tomamos el restaurante "subway" para valorar la correlación de este en cuanto al servicio con los demas establecimientos
me_gusta = "Subway"

names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

id_megusta

88

In [ ]:
corr_recomienda = corrMx[id_megusta]
print(corr_recomienda.shape)
print(corr_recomienda)

(129,)
[-2.91095724e-04 -3.60361693e-03  4.17542463e-03  5.49600644e-03
  8.22025821e-03  9.26584331e-04 -1.84324856e-04  2.08477911e-02
  4.16661248e-04  5.49213204e-03 -9.83321202e-03  2.30697829e-01
 -7.71002964e-04 -7.78964069e-03  4.28478573e-03 -1.96660815e-03
  6.46627016e-03  1.22858323e-02  5.23591750e-03 -3.37285475e-03
  4.22033518e-03  1.92759594e-01  3.25932626e-03  1.31225788e-02
  1.02856131e-02  1.08386707e-02 -2.32136875e-03  9.93522146e-03
 -8.13035667e-03  1.78740094e-01  1.27492522e-03  4.01749269e-03
  9.93395882e-03  1.06861306e-02  1.08193009e-02  6.49374214e-03
  2.57816306e-03  1.10159929e-02  1.62133585e-02  9.40455020e-03
  1.04115070e-02  1.39761855e-01  4.95429442e-03  1.01065382e-02
  7.45513647e-03  1.62758051e-02  3.65933684e-01 -4.53806898e-03
  1.94095851e-03  6.28735982e-03  7.08759527e-03  4.35778543e-03
  7.61149284e-03  6.86889327e-03  1.82869479e-01  2.26266242e-01
  8.41417946e-03 -1.38744406e-02  7.72973139e-01  2.58304203e-04
  6.00160625e-01  

In [ ]:
#Obtenemos el valor de la correlación del propio establecmiento para tener en cuenta el valor y no incluirlo en los mejores 10
corr_recomienda[id_megusta]

1.0

In [ ]:
print('Mejores 10 de acuerdo a correlación con restaurante "Subway" de acuerdo a la valoración del servicio\n')
mejores10=list(names[(corr_recomienda > .15) & (corr_recomienda < 0.999999)])
for pos in mejores10:
 id_otra = names_list.index(pos)
 print('Restaurante',pos, ' | valor correlación: ', round(corr_recomienda[id_otra],2))

Mejores 10 de acuerdo a correlación con restaurante "Subway" de acuerdo a la valoración del servicio

Restaurante Chilis Cuernavaca  | valor correlación:  0.23
Restaurante Giovannis  | valor correlación:  0.19
Restaurante Kiku Cuernavaca  | valor correlación:  0.18
Restaurante Paniroles  | valor correlación:  0.37
Restaurante Restaurant Familiar El Chino  | valor correlación:  0.18
Restaurante Restaurant Las Mananitas  | valor correlación:  0.23
Restaurante Restaurant Teely  | valor correlación:  0.77
Restaurante Restaurant and Bar and Clothesline Carlos N Charlies  | valor correlación:  0.6
Restaurante Rincon del Bife  | valor correlación:  0.52
Restaurante Vips  | valor correlación:  0.42



### Ejercicio 7.4

In [ ]:
num_sv = 12

SVD = TruncatedSVD(n_components=num_sv, random_state=1)

resultant_matrix7 = SVD.fit_transform(X)
resultant_matrix7.shape

me_gusta = "Subway"
names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

corrMx7 = np.corrcoef(resultant_matrix7, )
corr_recomienda7 = corrMx7[id_megusta]

id_megusta
corr_recomienda7[id_megusta]
Best10=list(names[(corr_recomienda7 > .9992) & (corr_recomienda7 < .999999)])
for pos in Best10:
 id_otra = names_list.index(pos)
 print('Restaurante',pos, ' | valor correlación: ', round(corr_recomienda7[id_otra],4))

Restaurante Chilis Cuernavaca  | valor correlación:  0.9999
Restaurante Kiku Cuernavaca  | valor correlación:  0.9993
Restaurante Log Yin  | valor correlación:  0.9999
Restaurante Los Vikingos  | valor correlación:  0.9993
Restaurante Mariscos Tia Licha  | valor correlación:  0.9993
Restaurante Paniroles  | valor correlación:  0.9998
Restaurante Restaurant Bar Coty y Pablo  | valor correlación:  0.9995
Restaurante Restaurant Familiar El Chino  | valor correlación:  0.9994
Restaurante Restaurant Las Mananitas  | valor correlación:  0.9992
Restaurante Restaurant Teely  | valor correlación:  0.9999
Restaurante Restaurant and Bar and Clothesline Carlos N Charlies  | valor correlación:  0.9999
Restaurante Rincon del Bife  | valor correlación:  1.0
Restaurante Sanborns Casa Piedra  | valor correlación:  0.9996
Restaurante Vips  | valor correlación:  0.9998


In [ ]:
print("Numero de Recomandaciones iguales:",len(set(mejores10).intersection(Best10)))

Numero de Recomandaciones iguales: 9


### Conclusiones

Después de realizado el ejercicio aplicando técnica de reducción de dimensionalidad para obtener recomendaciones de acuerdo a la valoración del servicio en los diferentes restaurantes, podemos observar que no existe mucha correlación entre los diferentes restaurantes en términos de la valoración indicada, podemos ver que al aplicar la técnica con todos los componentes obtenemos el ranking de los 10 mejores, donde solo el restaurante Teely tiene un coeficiente de correlación del 0.77 siendo el más alto, seguido de “Restaurant and Bar and Clothesline Carlos N Charlies” con un valor de 0.6 en la correlación, los que lo siguen tienen valores inferiores, por lo tanto al aplicar la técnica SVD encontrando un numero óptimo de 12 vectores singulares se logró obtener entre los 10 mejores 7 restaurantes que se obtuvieron con todos los componentes. De lo anterior podemos concluir que no existe una fuerte relación entre las valoraciones dadas por los usuarios en los diferentes restaurantes de acuerdo con el servicio.

### Conclusiones Finales

En los ejercicios realizados pudimos observar la importancia del método SVD en la creación de un modelo para sistemas de recomendación, ya que por medio de esta técnica podemos encontrar factores latentes para el modelo que nos permiten hacer predicciones con menor número de características, en la vida real podemos tener cantidad de columnas en nuestra matriz de utilidad que haría el procesamiento costoso en términos de tiempo y poder de cómputo, pero con el método pudimos probar que obteniendo un numero de factores mucho menor se puede lograr un resultado aproximadamente igual de eficiente que con todo el conjunto de datos. También pudimos observar que esta técnica nos permite hacer relaciones entre usuarios y los ítems relacionados, como fue el caso de los clientes y las valoraciones dadas a los restaurantes por tres aspectos diferentes, de esta forma se podrían modelar las características del sistema de recomendación. Como ultimo también se debe aclarar que como todo modelo se debe contar con la cantidad información suficiente para poder entrenar al modelo y que este tenga las suficientes bases para realizar las recomendaciones correctas de acuerdo con la realidad.